protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufRequestConverterFunction.java

```diff
@@ -0,0 +1,229 @@
+
+    private static final ConcurrentMap<Class<?>, MethodHandle> methodCache =
+            new MapMaker().weakKeys().makeMap();
+    private static final Parser defaultJsonParser = JsonFormat.parser().ignoringUnknownFields();
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    private static final MethodHandle unknownMethodHandle;
+
+    static {
+        final MethodHandles.Lookup publicLookup = MethodHandles.publicLookup();
+        final MethodType mt = methodType(void.class);
+
+        MethodHandle methodHandle;
+        try {
+            methodHandle = publicLookup.findConstructor(ProtobufRequestConverterFunction.class, mt);
+        } catch (NoSuchMethodException | IllegalAccessException e) {
+            methodHandle = null;
+        }
+        assert methodHandle != null;
+        unknownMethodHandle = methodHandle;
+    }
+
+    private final ExtensionRegistry extensionRegistry;
+    private final Parser jsonParser;
+    private final ResultType resultType;
+
+    ProtobufRequestConverterFunction(ResultType resultType) {
+        jsonParser = defaultJsonParser;
+        extensionRegistry = ExtensionRegistry.getEmptyRegistry();
+        this.resultType = requireNonNull(resultType, "resultType");
+    }
+
+    /**
+     * Creates an instance with the default {@link Parser} and {@link ExtensionRegistry}.
+     */
+    public ProtobufRequestConverterFunction() {
+        this(defaultJsonParser, ExtensionRegistry.getEmptyRegistry());
+    }
+
+    /**
+     * Creates an instance with the specified {@link Parser} and {@link ExtensionRegistry}.
+     */
+    public ProtobufRequestConverterFunction(Parser jsonParser, ExtensionRegistry extensionRegistry) {
+        this.jsonParser = requireNonNull(jsonParser, "jsonParser");
+        this.extensionRegistry = requireNonNull(extensionRegistry, "extensionRegistry");
+        resultType = ResultType.UNKNOWN;
+    }
+
+    @Nullable
+    @Override
+    public Object convertRequest(ServiceRequestContext ctx, AggregatedHttpRequest request,
+                                 Class<?> expectedResultType,
+                                 @Nullable ParameterizedType expectedParameterizedResultType) throws Exception {
+
+        final MediaType contentType = request.contentType();
+        if (resultType == ResultType.PROTOBUF ||
+            (resultType == ResultType.UNKNOWN && Message.class.isAssignableFrom(expectedResultType))) {
+            final Message.Builder messageBuilder = getMessageBuilder(expectedResultType);
+
+            if (contentType == null ||
+                contentType.subtype().contains("protobuf") || contentType.is(MediaType.OCTET_STREAM)) {
+                try (InputStream is = request.content().toInputStream()) {
+                    return messageBuilder.mergeFrom(is, extensionRegistry).build();
+                }
+            }
+
+            if (isJson(contentType)) {
+                jsonParser.merge(request.contentUtf8(), messageBuilder);
```

Comment:
Shouldn't we use charset?
`jsonParser.merge(request.content().toReader(contentType.charset(Charsets.UTF_8)), ...`

Reply:
I thought UTF-8 is only used for JSON.

Just out of curiosity, which is better `toReader(Charset)` or `content(Charset)` here?
If you are also curious, let me run a simple benchmark for it. 😉

Reply:
I think using the reader was more efficient but not 100% sure. 😄 
Could you run the benchmark if you don't mind?

Reply:
There is no big difference among them. `contentUtf8()` is 3% ~ 4% faster. It seems that within tolerance.

```
GrpcParseBenchmark.parseWithReader  thrpt    5  777.928 ± 4.739  ops/s
GrpcParseBenchmark.parseWithString  thrpt    5  806.419 ± 8.637  ops/s
```

The benchamrk code:
```java
static final String json = JsonFormat.printer().print(SEARCH_RESPONSE);
static final HttpData httpData = HttpData.wrap(json.getBytes(StandardCharsets.UTF_8));
static final Parser parser = JsonFormat.parser();

@Benchmark
public SearchResponse parseWithString() throws InvalidProtocolBufferException {
    final SearchResponse.Builder builder = SearchResponse.newBuilder();
    parser.merge(httpData.toStringUtf8(), builder);
    return builder.build();
}

@Benchmark
public SearchResponse parseWithReader() throws IOException {
    final SearchResponse.Builder builder = SearchResponse.newBuilder();
    parser.merge(httpData.toReader(StandardCharsets.UTF_8), builder);
    return builder.build();
}
```

Reply:
Thanks for the test, yeah using `contentUtf8` seems more efficient. 😄